Lucas match rating challenge/fix

jlamenha opened the pull request:
# This branch contains all of the changes that I have made to fix the average rating issue.

<b> I did use coding assistants, and I will include my prompts at the end of this for the sake of cleanliness. 
No coding assistants were used in writing this description. </b>

### The approach

#### This is the part that I am most proud of:

The approach I used to fix this bug involves storing the amount of times the player's running average was changed to ensure
that whenever it is changed, the amount of previous ratings is also taken into consideration. This is done through the following formula:

average = ((old_average * old_rating_count) + new_rating) / new_rating_count {or old_rating_count + 1}

This formula ensures that space is preserved in the storage to not have to store every single old rating that the player was given. Also saves computing costs by not running an average EVERY TIME an new rating is added (O(1) vs (O(N); where n is the amount of ratings a user has received, so it saves A LOT of computation as time goes on)

I realized soon into the challenge that I would not be able to change the API, as per the instructions. Because of this, I decided to use localStorage to make the changes that I needed to make persistent. This approach ensures that the data is saved locally despite reloading and recompiling the app (The only thing that should delete the data is running localStorage.clear(); on the console).

### The prompts

#### I'm not particularly proud of taking this approach, but it helped me save a lot of time and brain cells

I mostly used them for debugging issues I had with localStorage and some earlier approaches that I took, as well as to help me get started with the challenge.

To boot, I used this prompt to assist me with the initial problem I had, feeding it the base MatchRating.js file
![image](https://github.com/user-attachments/assets/d777d098-4c3c-4fd0-b29f-80d03e08a82d)

After this problem was solved and debugged, I moved on to a different problem that I noticed, which is where the running average formula came in. The prompt I used was:

"Okay, now I want to scale my ratings with running averages"

Once I figured this part out (which took no time), I ran into some issues with persisting averages, and then started using the localStorage to my advange, much like the base code does.

I stored an array of ratingCounts, which would help me store the amount of times each player's rating had been changed. I prompted my coding assistant to help me write some utils that I would use to store my data in the localStorage. 

For that, I used the following prompt:

"I want to keep a count for each player based on an array. E.g. [1,4,4] -> Player 1's count = 1, player 2's =4, and so on. Help me implement localStorage to ensure these changes are persistent across page refreshes."

After debugging, my implementation was finished.

jlamenha commented on code-challenge/src/components/MatchRating.js:
```diff
@@ -19,8 +21,33 @@ const MatchRating = ({ players, setPlayers }) => {
+
+   
+      // Do update the API, but this will not be accessed again (only able to store the last rating average with no running weights)
+      await submitRating(id, rating, players);
```
I see that now 🤦. I removed the await command.

jlamenha commented on code-challenge/src/utils/localStorageHelpers.js:
```diff
@@ -0,0 +1,28 @@
+const LS_KEY = "playerRatingState";
```
It is NOT, I am in fact, removing this file completely

jlamenha commented:
I updated my approach to this solution. I changed MatchRating.js to make more educated API calls to make the averaging calculation local. I wrote this in the file, but this way, we avoid using space in the user's device, but use more "networking cost." (compared to my older code, which did this calculation locally, while only storing the data locally as well.)

In MatchRating.js, I also mentioned that I removed the ratingState being passed down at App level, rather doing it on Component level. This makes App.js and PlayerList.js more readable (code-wise), while implementing the changes I mention earlier.

Looking back at my older code, it seems unproductive to only store the newer data locally without making use of the api that is available there. I am glad I got to come back to this and make a better solution.

jmusterman commented on code-challenge/src/components/MatchRating.js:
```diff
@@ -1,13 +1,19 @@
 import React, { useState } from 'react';
-import { submitRating } from '../api/ratingApi';
+// Changed the submitRating import to make more API calls directly
+import { storeRating, calculateAverageRating } from '../api/ratingApi';
```
yes!

jmusterman commented on code-challenge/src/components/MatchRating.js:
```diff
@@ -19,8 +25,41 @@ const MatchRating = ({ players, setPlayers }) => {
     
     try {
-      const updatedPlayers = await submitRating(selectedPlayer, rating, players);
+
```
this is a much better approach!  the main thing I was looking for is, while we do need to await the storeRating, almost everything else doesn't need to be awaited, it can actually be handled async.  This is commonly known as blocking vs non-blocking actions.  It is important to know what will directly impact the user's view and what can be processed in the background, especially as apps and APIs become more complex and can chain operations or have expensive reads/writes.

jmusterman commented on code-challenge/src/components/MatchRating.js:
```diff
@@ -19,8 +25,41 @@ const MatchRating = ({ players, setPlayers }) => {
+      await storeRating(id, rating);
+
+      // Calculate accurate average
+      const updatedAverage = await calculateAverageRating(id);
```
this probably doesn't need to be awaited.

jlamenha replied:
I ran into some UI issues when I removed this, but I explain more in a comment down below

jmusterman commented on code-challenge/src/components/MatchRating.js:
```diff
@@ -19,8 +25,41 @@ const MatchRating = ({ players, setPlayers }) => {
+      );
+
+      // Save updated list
+      await updatePlayers(updatedPlayers);
```
also probably doesn't need to be awaited

jmusterman reviewed (commented):
better!

jlamenha commented:
I made some changes. I removed the unnecessary awaits you mentioned, but I ran into some UI issues when I removed one of the calculateAverageRating awaits, so I decided to render an ellipsis in the rating tables for two reasons:

1. I wanted a visual indicator that the new average was being calculated that wasn't just the old average
2. It would break the app if I didn't do anything about it

(On a side note, I'm really enjoying making fixes to this code. :) )
